src/main/java/org/opensearch/knn/training/TrainingJob.java

```diff
@@ -0,0 +1,155 @@
+import org.opensearch.knn.indices.ModelState;
+
+import java.util.Objects;
+
+/**
+ * Encapsulates all information required to generate and train a model.
+ */
+public class TrainingJob implements Runnable {
+
+    public static Logger logger = LogManager.getLogger(TrainingJob.class);
+
+    private final KNNMethodContext knnMethodContext;
+    private final NativeMemoryCacheManager nativeMemoryCacheManager;
+    private final NativeMemoryEntryContext.TrainingDataEntryContext trainingDataEntryContext;
+    private final Model model;
+
+    private String modelId;
+
+    /**
+     * Constructor.
+     *
+     * @param modelId String to identify model. If null, one will be generated.
+     * @param knnMethodContext Method definition used to construct model.
+     * @param nativeMemoryCacheManager Cache manager loads training data into native memory.
+     * @param trainingDataEntryContext Training data configuration
+     * @param dimension model's dimension
+     * @param description user provided description of the model.
+     */
+    public TrainingJob(String modelId, KNNMethodContext knnMethodContext,
+                       NativeMemoryCacheManager nativeMemoryCacheManager,
+                       NativeMemoryEntryContext.TrainingDataEntryContext trainingDataEntryContext,
+                       int dimension, String description) {
+        this.modelId = modelId;
+        this.knnMethodContext = Objects.requireNonNull(knnMethodContext, "MethodContext cannot be null.");
+        this.nativeMemoryCacheManager = Objects.requireNonNull(nativeMemoryCacheManager,
+                "NativeMemoryCacheManager cannot be null.");
+        this.trainingDataEntryContext = Objects.requireNonNull(trainingDataEntryContext,
+                "TrainingDataEntryContext cannot be null.");
+        this.model = new Model(
+                        new ModelMetadata(
+                                knnMethodContext.getEngine(),
+                                knnMethodContext.getSpaceType(),
+                                dimension,
+                                ModelState.TRAINING,
+                                TimeValue.timeValueMillis(System.currentTimeMillis()),
+                                description,
+                                ""
+                        ),
+                        null
+                    );
+    }
+
+    /**
+     * Getter for model id.
+     *
+     * @return modelId
+     */
+    public String getModelId() {
+        return modelId;
+    }
+
+    /**
+     * Setter for model id.
+     *
+     * @param modelId to set
+     */
+    public void setModelId(String modelId) {
+        this.modelId = modelId;
+    }
+
+    /**
+     * Getter for model
+     *
+     * @return model
+     */
+    public Model getModel() {
+        return model;
+    }
+
+    @Override
+    public void run() {
+        NativeMemoryAllocation nativeMemoryAllocation = null;
+        ModelMetadata modelMetadata = model.getModelMetadata();
+
```

Comment:
Should we add a check to see if model is already in **created** state or **ModelBlob** is already present and then throw error? This can avoid cases of updating already existing model

Reply:
I intentionally structured this class not to interact with modelDao. In trainingJobRunner, it first attempts to put the initial model in the train state in modeldao. If a model already exists there, it will fail.